src/acl.c

```diff
@@ -1796,36 +1804,46 @@ void ACLFreeLogEntry(void *leptr) {
  *
  * The argpos argument is used when the reason is ACL_DENIED_KEY or 
  * ACL_DENIED_CHANNEL, since it allows the function to log the key or channel
- * name that caused the problem. Similarly the username is only passed when we
- * failed to authenticate the user with AUTH or HELLO, for the ACL_DENIED_AUTH
- * reason. Otherwise it will just be NULL.
+ * name that caused the problem.
+ *
+ * The last 3 arguments are a manual override to be used, instead of any of the automatic ones which
+ * depend on the c and reason arguments.
  */
-void addACLLogEntry(client *c, int reason, int argpos, sds username) {
+void addACLLogEntry(client *c, int reason, int argpos, sds username, sds object, int context) {
     /* Create a new entry. */
     struct ACLLogEntry *le = zmalloc(sizeof(*le));
     le->count = 1;
     le->reason = reason;
-    le->username = sdsdup(reason == ACL_DENIED_AUTH ? username : c->user->name);
+    le->username = sdsdup(username ? username : c->user->name);
     le->ctime = mstime();
 
-    switch(reason) {
-    case ACL_DENIED_CMD: le->object = sdsnew(c->cmd->name); break;
-    case ACL_DENIED_KEY: le->object = sdsdup(c->argv[argpos]->ptr); break;
-    case ACL_DENIED_CHANNEL: le->object = sdsdup(c->argv[argpos]->ptr); break;
-    case ACL_DENIED_AUTH: le->object = sdsdup(c->argv[0]->ptr); break;
-    default: le->object = sdsempty();
+    if (object) {
+        le->object = sdsnew(object);
+    } else {
+        switch(reason) {
+            case ACL_DENIED_CMD: le->object = sdsnew(c->cmd->name); break;
+            case ACL_DENIED_KEY: le->object = sdsdup(c->argv[argpos]->ptr); break;
+            case ACL_DENIED_CHANNEL: le->object = sdsdup(c->argv[argpos]->ptr); break;
+            case ACL_DENIED_AUTH: le->object = sdsdup(c->argv[0]->ptr); break;
+            default: le->object = sdsempty();
+        }
     }
 
     client *realclient = c;
     if (realclient->flags & CLIENT_LUA) realclient = server.lua_caller;
 
     le->cinfo = catClientInfoString(sdsempty(),realclient);
-    if (c->flags & CLIENT_MULTI) {
-        le->context = ACL_LOG_CTX_MULTI;
-    } else if (c->flags & CLIENT_LUA) {
-        le->context = ACL_LOG_CTX_LUA;
+
+    if (context != -1) {
```

Comment:
The context is an enum value, can we just use a defined value instead of passing in -1, which is so cryptic?

Also, I'm not clear why we don't just pass the context in, since I assume we know it better from the calling site. This seems like an unnecessary level of abstraction.

Reply:
We actually don't know the context in the caller (can't be hard coded there), and Salvatore designed all of these to be automatically populated by the server struct state. Maybe we can somehow do that for modules too?

Reply:
We actually can, we call it from processCommand (top_context), execCommand (multi_context) and luaRedisGenericCommand (lua_context)

Reply:
I think it makes sense to take it out of `addACLLogEntry`. 

BTW I find the context pretty confusing - what should I expect if I have an `EVAL` calling a module command all wrapped in `MULTI/EXEC`?